CONTRIBUTING.md

```diff
@@ -23,28 +23,49 @@ with the [latest beta of rclone](https://beta.rclone.org/):
 If you find a bug that you'd like to fix, or a new feature that you'd
 like to implement then please submit a pull request via GitHub.
 
-If it is a big feature then make an issue first so it can be discussed.
-
-You'll need a Go environment set up with GOPATH set.  See [the Go
-getting started docs](https://golang.org/doc/install) for more info.
+If it is a big feature, then [make an issue](https://github.com/rclone/rclone/issues) first so it can be discussed.
 
 First in your web browser press the fork button on [rclone's GitHub
 page](https://github.com/rclone/rclone).
 
-Now in your terminal
+Then [install Git](https://git-scm.com/downloads) and set your public contribution [name](https://docs.github.com/en/github/getting-started-with-github/setting-your-username-in-git) and [email](https://docs.github.com/en/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address#setting-your-commit-email-address-in-git).
+
+Next open your terminal, change directory to your preferred folder and  initialize your local rclone project:
 
     git clone https://github.com/rclone/rclone.git
     cd rclone
     git remote rename origin upstream
+      # if you have SSH keys setup in your GitHub account:
     git remote add origin git@github.com:YOURUSER/rclone.git
-    go build
+      # otherwise:
+    git remote add origin https://github.com/YOURUSER/rclone.git
+
+Note that most of the terminal commands in the rest of this guide must be executed from the rclone folder created above.
+
+Now [install Go](https://golang.org/doc/install) and verify the [GOPATH](https://github.com/golang/go/wiki/GOPATH) is pointing to a suitable folder:
+
+    go env GOPATH
+
+Great, you can now compile and execute your own version of rclone:
+
+    go build                              
+    rclone version
+
+Finally update your forked code to the latest version of the [rclone master](https://github.com/rclone/rclone/tree/master):
+
+    git checkout master
+    git fetch upstream
```

Comment:
I'd normally recommend `git pull --rebase upstream` for Git newbies here instead of

    git fetch upstream
    git merge --ff-only

Especially if you've been working on master.

I guess `git pull --rebase` potentially leaves you sorting out a conflict whereas `git merge --ff-only` will refuse the merge.

Reply:
As a beginner I really like the transparency of first fetching from remote to local repo, then comparing local repo to my workspace and then merging the local repo to my workspace. It helped me tremendously in building my conceptual model of Git.

I was utterly confused when I first read about git pull before understanding the basic principles of fetch, merge and the difference between remote, local and working space. Just too much in one mouthful for me.

I therefore intentionally split the commands to ease the Git understanding for curious contributors googling the commands before execution. I guess experienced Gitters will just do a corresponding pull of their own liking.

You are right, `–ff-only` will refuse the merge if you accidently worked in master. In this situation I prefer to fail-fast (and hard). I haven’t tried, but I assume the mess is smaller and therefore easier to clean-up compared to the situation after a (series of) rebases (with or without conflicts). 

I’m also worried that (thoughtless) rebasing could lead to issues in later work. Suppose a beginner overlooks a misplaced commit in a successfully rebased master. Couldn’t that lead to troubles sometime later?

These were the (possibly naive) thoughts behind my proposal. You are the one with the experience in Git and the typical needs/mistakes of contributors, so I will leave the decision to you.

Reply:
I think your proposal is a good one. When I cast my mind back to when I was a git beginner it was having to deal with merge conflicts that caused me to do stuff like delete the whole repo and start again ;-)

> I’m also worried that (thoughtless) rebasing could lead to issues in later work. Suppose a beginner overlooks a misplaced commit in a successfully rebased master. Couldn’t that lead to troubles sometime later?

Rebasing either works, or gives a conflict. You can think of a rebase as saving your existing work as patches then reapplying them to the branch you are rebasing on. If there is a conflict in those patches you'll have to fix it.

So rebasing  should be harmless and any branches will need to be rebased eventually since rclone uses a [linear history model with no merges](https://www.bitsnbites.eu/a-tidy-linear-git-history/). The GithHub merge PR interface will do the rebase for me if there are no conflicts.

Reply:
Thanks, and nice to learn more about rebasing, linear history and GitHub.

I still haven’t tried a merge conflict and do my very best to avoid it. I should probably try it out one day on a test branch to get more comfortable.